### PR TITLE
Refactor SLI test for spike generator to Python

### PR DESCRIPTION
**Pull request by pnbabu**

### CODE REVIEW

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    sg.set(sg_params)
+
+    sr = nest.Create("spike_recorder")
+    sr.set({"time_in_steps": True})
+    nest.Connect(sg, sr, syn_spec={"delay": 1.0, "weight": 1.0})
+
+    nest.Simulate(10.)
+
+    spike_times = sr.get("events")["times"]
+    nptest.assert_array_equal(spike_times, expected_spike_times)
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:77]:
```suggestion
def test_spike_generator(prepare_kernel, spike_times, allow_offgrid_times, expected_spike_times):
    sg_params = {
        "precise_times": False,
        "spike_times": spike_times,
        "allow_offgrid_times": allow_offgrid_times,
        "origin": 0.0,
        "start": 0.0,
        "stop": 6.0,
    }

    sg = nest.Create("spike_generator", sg_params)
    sr = nest.Create("spike_recorder", {"time_in_steps": True})
    nest.Connect(sg, sr, syn_spec={"delay": 1.0, "weight": 1.0})

    nest.Simulate(10.0)

    actual_spike_times = sr.events["times"]
    nptest.assert_array_equal(actual_spike_times, expected_spike_times)
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                             [{"spike_times": [0.0, 10.0, 10.5, 10.50001], "shift_now_spikes": True},
+                              [0.1, 10.0, 10.5, 10.5]]
+                         ])
+def test_spike_generator_set_and_get(prepare_kernel, sg_params, expected_spike_times):
+    sg = nest.Create("spike_generator")
+    sg.set(sg_params)
+
+    # Get spike times
+    spike_times = sg.get("spike_times")
+    assert spike_times == pytest.approx(expected_spike_times)
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:211]:
```suggestion
def test_spike_generator_set_and_get(prepare_kernel, sg_params, expected_spike_times):
    sg = nest.Create("spike_generator")
    sg.set(sg_params)
    actual_spike_times = sg.get("spike_times")
    nptest.assert_almost_equal(actual_spike_times, expected_spike_times, decimal=5)
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                 "stop": 6.}
+    sg.set(sg_params)
+
+    sr = nest.Create("spike_recorder")
+    nest.Connect(sg, sr, syn_spec={"delay": 1.0, "weight": 1.0})
+
+    nest.Simulate(7.0)
+
+    spike_times = sr.get("events")["times"]
+    assert spike_times == pytest.approx(expected_spike_times)
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:233]:
```suggestion
def test_spike_generator_precise_times_different_resolution(h, expected_spike_times):
    nest.ResetKernel()
    nest.resolution = h

    sg_params = {
        "precise_times": True,
        "spike_times": [0.1, 5.0, 5.3, 5.300001, 5.399999, 5.9, 6.0, 9.3],
        "origin": 0.0,
        "start": 5.0,
        "stop": 6.0,
    }
    sg = nest.Create("spike_generator", sg_params)
    sr = nest.Create("spike_recorder")
    nest.Connect(sg, sr, syn_spec={"delay": 1.0, "weight": 1.0})

    nest.Simulate(7.0)

    actual_spike_times = sr.events["times"]
    nptest.assert_almost_equal(actual_spike_times, expected_spike_times)
```

**Review by nicolossus — CHANGES_REQUESTED** [review]
@pnbabu Thanks, this looks good overall. I've made some suggestions for making the test code a bit more compact and formatted according to the Black formatter. 

The SLI versions of these tests have some documentation; can you please also add docstrings with a short descriptions for each test?

**Comment by heplesser** [comment]
@pnbabu Ping from the NEST Testing meeting!

**Comment by heplesser** [comment]
@pnbabu Ping!

**Comment by pnbabu** [comment]
> @pnbabu Ping!

I have applied the suggestions from the code review. I will request a review again.

**Review by nicolossus — CHANGES_REQUESTED** [review]
@pnbabu Thanks! This now looks mostly good to me, but can you also add the documentation from the SLI version as docstrings describing a bit more what each test actually checks?

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    actual_offsets = sr.events["offsets"]
+    nptest.assert_almost_equal(actual_offsets, expected_offsets)
+
+
+def test_spike_generator_spike_time_at_simulation_end_time(prepare_kernel):
+    """
+    This test checks the spikes after the simulation time are not recorded.
+
+    Here, the spike time is within tic/2 of step 100, rounded down to 100 thus not in the future; spike will not
+    be emitted
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:136]:
```suggestion
    Here, the spike time is within ``tic/2`` of step 100, rounded down to 100 thus not in the future;
    spike will not be emitted.
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    sg.set(sg_params)
+
+    nest.Simulate(10.0)
+
+    assert sr.n_events == 0
+
+
+def test_spike_generator_precise_time_future_spike(prepare_kernel):
+    """
+    This test checks the future spike times with "precise_times" set to True.
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:154]:
```suggestion
    Ensure correct behavior of future spike times with ``precise_times`` set to ``True``.
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    nptest.assert_array_equal(actual_spike_times, expected_spike_times)
+
+    expected_offsets = [0.0999]
+    actual_offsets = sr.events["offsets"]
+    nptest.assert_almost_equal(actual_offsets, expected_offsets)
+
+
+def test_spike_generator_with_shift_now_spikes(prepare_kernel):
+    """
+    This test checks the spike times with "shift_now_spikes" set to True.
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:180]:
```suggestion
    Check spike times with ``shift_now_spikes`` set to ``True``.
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        nest.kernel.NESTError,
+        match="Option precise_times cannot be set to true when either allow_offgrid_times or "
+        "shift_now_spikes is set to true.",
+    ):
+        sg = nest.Create("spike_generator", {"precise_times": True, "allow_offgrid_times": True})
+
+
+def test_spike_generator_precise_times_and_shift_now_spikes(prepare_kernel):
+    """
+    This test ensures the exclusivity between options "precise_times" and "shift_now_spikes".
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:219]:
```suggestion
    Ensure exclusivity between options ``precise_times`` and ``shift_now_spikes``.
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        [{"spike_times": [0.0, 10.0, 10.5, 10.50001], "shift_now_spikes": True}, [0.1, 10.0, 10.5, 10.5]],
+    ],
+)
+def test_spike_generator_set_and_get(prepare_kernel, sg_params, expected_spike_times):
+    """
+    This test checks the set and get functions of the spike generator.
+    - set and get with default values
+    - set and get with "allow_offgrid_times" set to True
+    - set and get with "precise_times" set to True
+    - set and get with "shift_now_spikes" set to True
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:247]:
```suggestion
    Test the ``set`` and ``get`` functions of ``spike_generator``.
    
    The test checks ``set`` and ``get`` with:
    - default values.
    - ``allow_offgrid_times`` set to ``True``.
    - ``precise_times`` set to ``True``.
    - ``shift_now_spikes`` set to ``True``.
```

File: testsuite/pytests/sli2py_stimulating/test_spike_generator.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        0.1,
+        0.2,
+        0.5,
+        1.0,
+    ],
+)
+@pytest.mark.parametrize("expected_spike_times", [[5.3, 5.300001, 5.399999, 5.9, 6.0]])
+def test_spike_generator_precise_times_different_resolution(h, expected_spike_times):
+    """
+    This test checks the precise times of spikes for different simulation resolution values.
```
Comment (nicolossus) [inline testsuite/pytests/sli2py_stimulating/test_spike_generator.py:267]:
```suggestion
    Test the precise times of spikes for different resolutions.
```

**Review by nicolossus — CHANGES_REQUESTED** [review]
@pnbabu Thanks! I have a few nitpicks but once they are addressed I will be happy :)

**Review by nicolossus — APPROVED** [review]

**Pull request merged**